src/DidCache.ts

```diff
@@ -1,16 +1,338 @@
+    const op = await this.getOperation(opInfo);
+
+    if (this.isInitialVersion(opInfo)) {
+      return this.didDocCreate(op);
+    } else {
+      const prevVersion = op.previousOperationHash as VersionId;
+      const prevDidDoc = await this.lookup(prevVersion);
+      if (prevDidDoc === null) {
+        return null;
+      } else {
+        return this.didDocUpdate(prevDidDoc, op);
+      }
+    }
+  }
+
+  /**
+   * Return the previous version id of a given DID version. The implementation
+   * is inefficient and involves an async cas read. This should not be a problem
+   * since this method is not hit for any of the externally exposed DID operations.
+   */
+  public async prev (versionId: VersionId): Promise<VersionId | null> {
+    const opInfo = this.opHashToInfo.get(versionId);
+    if (opInfo) {
+      const op = await this.getOperation(opInfo);
+      if (op.previousOperationHash) {
+        return op.previousOperationHash;
+      }
+    }
+    return null;
+  }
+
+  /**
+   * Return the first version of a DID document given a possibly later version.
+   * A simple recursive implementation using prev; not very efficient but should
+   * not matter since this method is not hit for any externally exposed DID
+   * operations.
+   */
+  public async first (versionId: VersionId): Promise<VersionId | null> {
+    const opInfo = this.opHashToInfo.get(versionId);
+    if (opInfo === undefined) {
+      return null;
+    }
+
+    if (this.isInitialVersion(opInfo)) {
+      return versionId;
+    }
+
+    const prevVersionId = await this.prev(versionId);
+    if (prevVersionId === null) {
+      return null;
+    }
+
+    return this.first(prevVersionId);
+  }
+
+  /**
+   * Return the next version of a DID document if it exists or null, otherwise.
+   */
+  public next (versionId: VersionId): VersionId | null {
+    const nextVersionId = this.nextVersion.get(versionId);
+    if (nextVersionId === undefined) {
+      return null;
+    } else {
+      return nextVersionId;
+    }
+  }
+
+  /**
+   * Return the latest (most recent) version of a DID document. Return null if
+   * the version is unknown.
+   */
+  public last (versionId: VersionId): VersionId | null {
+    const opInfo = this.opHashToInfo.get(versionId);
+    if (opInfo === undefined) {
+      return null;
+    }
+
+    const nextVersionId = this.next(versionId);
+    if (nextVersionId === null) {
+      return versionId;
+    } else {
+      return this.last(nextVersionId);
+    }
+  }
+
+  /**
+   * Get the last processed transaction.
+   * TODO: fix this after discussing the intended semantics.
+   */
+  public get lastProcessedTransaction (): Transaction | undefined {
+    return undefined;
+  }
+
+  /**
+   * Get a cryptographic hash of the write operation. Currently, uses
+   * SHA256 to get hashes (TODO: Fix it to be consistent DID generation)
+   */
+  private static getHash (operation: WriteOperation): OperationHash {
+    const sha256HashCode = 18;
```

Comment:
Can't hard-code hash function to use, has to depend on protocol version ie. Protocol.getProtocol(...), this means either the operationHash or blockNumber needs to be given in the WriteOperation instances. Again, I personally think operationHash would do, and thus no need to return the hash again when apply() is invoked.

Reply:
I don't understand your suggestion. Even if apply() does not return anything, the did cache needs to match previousOperationHash in an update operation with a previous operation that has this hash. This means that the cache should either (1) have the ability to compute hashes of operations or (2) the caller (e.g., observer) needs to provide the operation hash as part of an Operation. 

If we go with (1) we need to provide the blockNumber in WriteOperation; if we go with (2) we need to provide operationHash in WriteOperation. I don't have a strong preference between (1) and (2).

Reply:
We are in agreement, creating an issue to track this issue.